src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -47,21 +50,38 @@ public sealed class RestoreManagerPackage : AsyncPackage
         // won't get disconnected.
         private EnvDTE.BuildEvents _buildEvents;
 
+        // keep a reference to initialize task so that it can complete
+        // even after InitializeAsync() is done.
+        private JoinableTask _initializationTask;
+
         protected override async Task InitializeAsync(
             CancellationToken cancellationToken, 
             IProgress<ServiceProgressData> progress)
         {
             var componentModel = await GetServiceAsync(typeof(SComponentModel)) as IComponentModel;
             componentModel.DefaultCompositionService.SatisfyImportsOnce(this);
 
-            // Accessing DTE without confirming to Main thread because AllowsBackgroundLoading is false which
+            // Start this initialization task but don't await since it will degrade performance.
             // will make sure that this piece is always executed on Main thread.
-            var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
-            _buildEvents = dte.Events.BuildEvents;
-            _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
-
-            UserAgent.SetUserAgentString(
-                new UserAgentStringBuilder().WithVisualStudioSKU(dte.GetFullVsVersionString()));
+            // Besides we don't need to wait here to complete since this will only bind to build event which
+            _initializationTask = ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                try
+                {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var dte = (EnvDTE.DTE)await GetServiceAsync(typeof(SDTE));
+                    _buildEvents = dte.Events.BuildEvents;
+                    _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
+
+                    UserAgent.SetUserAgentString(
+                        new UserAgentStringBuilder().WithVisualStudioSKU(dte.GetFullVsVersionString()));
+                }
+                catch (Exception ex)
```

Comment:
Didn't notice this wasn't blocking at first since there is an await on another initialize below it. I don't think we should launch this without waiting. There is nothing stopping a build from starting before NuGet starts reading it except for pure luck. Potentially 100s of builds could take place before this thread wakes up and hooks up the event. VS automation is like to hit issues with this first.